library/chemistry/src/JordanWigner/StatePreparation.qs

```diff
@@ -0,0 +1,779 @@
+
+    mutable success = false;
+    repeat {
+        use auxillary = Qubit[nBitsIndices + 1];
+        use flag = Qubit();
+
+        let arr = Mapped(PrepareSingleConfigurationalStateSingleSiteOccupationWrapper, applyFlips);
+        let multiplexer = MultiplexerBruteForceFromGenerator(nExcitations, idx -> arr[idx]);
+        PrepareArbitraryStateCP(coefficientsNewComplexPolar, auxillary);
+        multiplexer(auxillary, qubits);
+        Adjoint PrepareArbitraryStateD(coefficientsSqrtAbs, auxillary);
+        ApplyControlledOnInt(0, X, auxillary, flag);
+
+        // if measurement outcome one we prepared required state
+        let outcome = M(flag);
+        success = outcome == One;
+        ResetAll(auxillary);
+        Reset(flag);
+    } until success
+    fixup {
+        ResetAll(qubits);
+    }
+}
+
+/// # Summary
+/// Given a set of coefficients and a little-endian encoded quantum register,
+/// prepares an state on that register described by the given coefficients.
+///
+/// # Description
+/// This operation prepares an arbitrary quantum
+/// state $\ket{\psi}$ with complex coefficients $r_j e^{i t_j}$ from
+/// the $n$-qubit computational basis state $\ket{0 \cdots 0}$.
+/// In particular, the action of this operation can be simulated by the
+/// a unitary transformation $U$ which acts on the all-zeros state as
+///
+/// $$
+/// \begin{align}
+///     U\ket{0...0}
+///         & = \ket{\psi} \\\\
+///         & = \frac{
+///                 \sum_{j=0}^{2^n-1} r_j e^{i t_j} \ket{j}
+///             }{
+///                 \sqrt{\sum_{j=0}^{2^n-1} |r_j|^2}
+///             }.
+/// \end{align}
+/// $$
+///
+/// # Input
+/// ## coefficients
+/// Array of up to $2^n$ complex coefficients represented by their
+/// absolute value and phase $(r_j, t_j)$. The $j$th coefficient
+/// indexes the number state $\ket{j}$ encoded in little-endian format.
+///
+/// ## qubits
+/// Qubit register encoding number states in little-endian format. This is
+/// expected to be initialized in the computational basis state
+/// $\ket{0...0}$.
+///
+/// # Remarks
+/// Negative input coefficients $r_j < 0$ will be treated as though
+/// positive with value $|r_j|$. `coefficients` will be padded with
+/// elements $(r_j, t_j) = (0.0, 0.0)$ if fewer than $2^n$ are
+/// specified.
+///
+/// # Example
+/// The following snippet prepares the quantum state $\ket{\psi}=\sqrt{1/8}\ket{0}+\sqrt{7/8}\ket{2}$
+/// in the qubit register `qubitsLE`.
+/// ```qsharp
+/// use qubits = Qubit();
+/// let qubitsLE = LittleEndian([qubits]);
+/// PrepareArbitraryStateCP([ComplexPolar(1.0/Sqrt(2.0),0.0),ComplexPolar(1.0/Sqrt(2.0),PI()/2.0)],qubitsLE); // = |i>
+/// ```
+///
+/// # References
+/// - [Synthesis of Quantum Logic Circuits](https://arxiv.org/abs/quant-ph/0406176)
+///   Vivek V. Shende, Stephen S. Bullock, Igor L. Markov
+operation PrepareArbitraryStateCP(coefficients : ComplexPolar[], qubits : Qubit[]) : Unit is Adj + Ctl {
```

Comment:
I'm removing some of these and replacing them with our Std library implementations.